homeassistant/components/integration/sensor.py

```diff
@@ -292,109 +383,50 @@ async def async_added_to_hass(self) -> None:
             self._attr_device_class = state.attributes.get(ATTR_DEVICE_CLASS)
             self._unit_of_measurement = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
 
-        @callback
-        def calc_integration(event: Event[EventStateChangedData]) -> None:
-            """Handle the sensor state changes."""
-            old_state = event.data["old_state"]
-            new_state = event.data["new_state"]
-
-            if (
-                source_state := self.hass.states.get(self._sensor_source_id)
-            ) is None or source_state.state == STATE_UNAVAILABLE:
-                self._attr_available = False
-                self.async_write_ha_state()
-                return
-
-            self._attr_available = True
-
-            if old_state is None or new_state is None:
-                # we can't calculate the elapsed time, so we can't calculate the integral
-                return
-
-            unit = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-            if unit is not None:
-                self._unit_of_measurement = self._unit(unit)
-
-            if (
-                self.device_class is None
-                and new_state.attributes.get(ATTR_DEVICE_CLASS)
-                == SensorDeviceClass.POWER
-            ):
-                self._attr_device_class = SensorDeviceClass.ENERGY
-                self._attr_icon = None
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass,
+                [self._sensor_source_id],
+                self._integrate_on_state_change_callback,
+            )
+        )
+
+    @callback
+    def _integrate_on_state_change_callback(
```

Comment:
The method name is misleading because it doesn't just calculate the integration, it updates the state including attributes